update proposal to use new access syntax v[A] (#42)
onflow/flips: cadence/2022-09-21-attachments.md

and fields of `I`. Note that an attachment must implement all the methods and fields of `I` itself; it may reference its base type with `super` but it cannot inherit 
from that type. 

Important note: because attachments are not first class values in Cadence, their types cannot appear outside of a reference type. So, for example, given an 

> attachment declaration `attachment A for X {}`, `A`, `A?`, `[A]` and `((): A)` are not valid type annotations, while `&A`, `A?`, `[&A]` and `((): &A)` are valid. 

Reviewer: ```suggestion
attachment declaration `attachment A for X {}`, the types `A`, `A?`, `[A]` and `((): A)` are not valid type annotations, while `&A`, `&A?`, `[&A]` and `((): &A)` are valid. 
```

> For this reason as well, `self` inside an attachment has a reference type; in `A` above, the type of self would be `&A` rather than `A`. 

Reviewer: ```suggestion
For this reason as well, `self` inside an attachment has a reference type. 
In the attachment declaration `A` above, the type of `self` would be `&A`, rather than `A` like in other composite declarations.
```

### Accessing Attachments

Once an attachment has been added to a composite value, it can be accessed using indexing syntax: `v[T]`, where `v` is a resource or struct composite value, 

> and `T` is the name of an attachment type. This indexing syntax returns an optional reference to the attachment of type `T` (`&T?`); if no such attachment exists on `v`, 

Reviewer: ```suggestion
and `T` is the name of an attachment type. This indexing syntax returns an optional reference to the attachment of type `T`: `&T?`; if no such attachment exists on `v`, 
```

the expression will return `nil`. So, given a composite `r` with an attachment of type `A`, accessing `A`'s `foo` method would be done like so:

```cadence
r[A]!.foo()
```

### Iterating over Attachments

All composite types will contain a new function `forEachAttachment` that iterates over all the attachments present on that composite. On a resource, this function 
will have the following signature:

```cadence
fun forEachAttachment(_ f: ((&AnyResourceAttachment): Void)): Void 
```

On a structure, it will have 

```cadence
fun forEachAttachment(_ f: ((&AnyStructAttachment): Void)): Void 
```

> `AnyResourceAttachment`/`AnyStructAttachment` are new types that expresses the supertype of all attachments, 

Reviewer: What was the reasoning again for having one supertype each. Maybe add an explanation why that is necessary.

> `AnyResourceAttachment`/`AnyStructAttachment` are new types that expresses the supertype of all attachments, 
> and contains two methods (that are implicitly present on all attachments): `getField` and `getMethod`, 

Reviewer: We consistently call functions on composites functions, and do not use the term "method" anywhere:

```suggestion
`AnyResourceAttachment`/`AnyStructAttachment` are new types that expresses the supertypes of all struct/resource attachments.
They contains two functions (that are implicitly present on all attachments): `getField` and `getFunction`, 
```
